sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/RegionScopedSessionContainer.java

```diff
@@ -0,0 +1,621 @@
+            this.collectionResourceIdToPartitionScopedRegionLevelProgress.compute(
+                resourceId.getUniqueDocumentCollectionId(), (k, partitionScopedRegionLevelProgressAsVal) -> {
+
+                    if (partitionScopedRegionLevelProgressAsVal == null) {
+                        logger.info("Registering a new collection resourceId [{}] in "
+                            + "RegionScopedSessionContainer", resourceId);
+                        partitionScopedRegionLevelProgressAsVal =
+                            new PartitionScopedRegionLevelProgress();
+                    }
+
+                    return partitionScopedRegionLevelProgressAsVal;
+                }
+            );
+
+            partitionScopedRegionLevelProgress =
+                this.collectionResourceIdToPartitionScopedRegionLevelProgress.get(resourceId.getUniqueDocumentCollectionId());
+
+            if (partitionScopedRegionLevelProgress != null) {
+                this.recordRegionScopedSessionToken(
+                    request,
+                    partitionScopedRegionLevelProgress,
+                    parsedSessionToken,
+                    partitionKeyRangeId,
+                    regionRoutedTo);
+            }
+
+            if (shouldUseBloomFilter(
+                request,
+                partitionKeyInternal,
+                partitionKeyDefinition)) {
+
+                this.recordPartitionKeyInBloomFilter(
+                    request,
+                    collectionResourceId,
+                    regionRoutedTo,
+                    partitionKeyInternal.v,
+                    partitionKeyDefinition.v);
+            }
+        }
+    }
+
+    private String getCombinedSessionToken(PartitionScopedRegionLevelProgress partitionScopedRegionLevelProgress) {
+        ConcurrentHashMap<String, ConcurrentHashMap<String, PartitionScopedRegionLevelProgress.RegionLevelProgress>> tokens
+            = partitionScopedRegionLevelProgress.getPartitionKeyRangeIdToRegionLevelProgress();
+
+        StringBuilder result = new StringBuilder();
+        if (tokens != null) {
+            for (Iterator<Map.Entry<String, ConcurrentHashMap<String, PartitionScopedRegionLevelProgress.RegionLevelProgress>>> iterator = tokens.entrySet().iterator(); iterator.hasNext(); ) {
+                Map.Entry<String,  ConcurrentHashMap<String, PartitionScopedRegionLevelProgress.RegionLevelProgress>> entry = iterator.next();
+
+                String partitionKeyRangeId = entry.getKey();
+                String sessionTokenAsString = entry.getValue().get(PartitionScopedRegionLevelProgress.GLOBAL_PROGRESS_KEY).getSessionToken().convertToString();
+
+                result = result.append(partitionKeyRangeId).append(":").append(sessionTokenAsString);
+                if (iterator.hasNext()) {
+                    result = result.append(",");
+                }
+            }
+        }
+
+        return result.toString();
+    }
+
+    private boolean shouldUpdateSessionToken(
```

Comment:
I guess we can also have a sessionContainer base class for common methods between the two implementations

Reply:
Added a `SessionContainerUtil` for the sharing common methods.